Setup: test, hmr

msuntharesan opened the pull request:
This is the initial commit of what I think are base setup for the dev
test
- karma
- tape

dev
- express
- webpack-dev-middleware

and some npm scripts

Commit:
Setup: test, hmr

test
 - karma
 - tape
dev
 - express
 - webpack-dev-middleware

and some npm scripts

Commit:
Make Travis happy

Commit:
Travis not happy yet

jnwng commented:
@vanthiyathevan first off, thank you for sending this through! its important to get things started somewhere. i'm going to go through and leave a few comments on the files / code submitted.

a few things i won't mention in the diff - we'll be following the [airbnb javascript styleguide](https://github.com/airbnb/javascript) and eslint will be modified accordingly. in the meantime, it would be useful to conform to style conventions (consistent quotes, trailing newlines) as much as possible.

jnwng commented on jsconfig.json:
```diff
@@ -0,0 +1,6 @@
+{
```
is this a `vscode` specific file? editor-specific configurations probably do not need to be included in the mainline repository

msuntharesan replied:
It is and has been removed.
But I think we should add the [editorconfig](http://editorconfig.org/)

jnwng commented on src/main.test.jsx:
```diff
@@ -0,0 +1,12 @@
+
+test('Main component', (t) => {
+  var root = TestUtils.renderIntoDocument(<Main/>);
+  debugger
```
this should be removed

jnwng commented on src/main.test.jsx:
```diff
@@ -0,0 +1,12 @@
+test('Main component', (t) => {
+  var root = TestUtils.renderIntoDocument(<Main/>);
+  debugger
+  t.ok(root, 'Componenet is ok');
```
Componenet -> Component

jnwng commented on webpack.config.js:
```diff
@@ -0,0 +1,65 @@
+      },
+      {
+        test: /(\.s?css)$/,
+        loader: ExtractTextPlugin.extract('style', 'css-loader?sourceMap&modules&importLoaders=1&localIdentName=[name]__[local]___[hash:base64:5]!postcss!sass?sourceMap=true&sourceMapContents=true!toolbox')
```
its much clearer to express this as a set of values passed via the `query` key in the [loader configuration](https://webpack.github.io/docs/loaders.html#loaders)

jnwng commented on webpack.config.test.js:
```diff
@@ -0,0 +1,40 @@
+        test: /(\.jsx?)$/,
+        exclude: /node_modules/,
+        loader: 'react-hot!babel',
+        presets: ['es2015', 'react', 'stage-0']
```
these can be better expressed via  [`.babelrc`](https://babeljs.io/docs/usage/babelrc/) file

jnwng commented on package.json:
```diff
@@ -44,16 +46,38 @@
+    "node-sass": "^3.4.2",
+    "phantomjs": "^1.9.19",
+    "postcss-loader": "^0.8.0",
+    "react-hot-loader": "^1.3.0",
```
[`react-hot-loader`](https://github.com/gaearon/react-hot-loader) has been deprecated in lieu of [babel-plugin-react-transform](https://www.npmjs.com/package/babel-plugin-react-transform) and [react-transform-hmr](https://www.npmjs.com/package/react-transform-hmr), and is the preferred solution for hotloading moving forward.

msuntharesan replied:
@jnwng [babel-plugin-react-transform](https://github.com/gaearon/babel-plugin-react-transform/pull/50) doesn't support babel6 yet :disappointed: 
We can switch over once its out

zaggy replied:
I would prefer to stick to babel 5 then until a babel6 support for babel-plugin-react-transform is out.

timaschew replied:
agree to use babel-plugin-react-transform and stick to version 5 for now

jnwng replied:
yep, another +1 here, the benefits of react-transform and the [class of plugins](https://www.npmjs.com/search?q=react-transform) it provides outweigh the benefits of babel 6 right now.

looks like babel6 support for react-transform is [imminent](https://github.com/gaearon/babel-plugin-react-transform/pull/50) anyway

Commit:
Linting:

- comply with eslint and airbnb
- remove unncessary files
- webpack config

abd124abd commented:
@jnwng - just ooc any notable mention(s) why airbnb linting/styling be the standard?

sokra commented on webpack.config.test.js:
```diff
@@ -0,0 +1,39 @@
+      },
+      {
+        test: /(\.s?css)$/,
+        loader: "css-loader?sourceMap&modules&importLoaders=1&localIdentName=[name]__[local]___[hash:base64:5]!postcss!sass?sourceMap=true&sourceMapContents=true",
```
`importLoaders` should be 2: `postcss` and `sass`

sokra commented on webpack.config.js:
```diff
@@ -0,0 +1,64 @@
+      },
+      {
+        test: /(\.s?css)$/,
+        loader: ExtractTextPlugin.extract("style", "css-loader?sourceMap&modules&importLoaders=1&localIdentName=[name]__[local]___[hash:base64:5]!postcss!sass?sourceMap=true&sourceMapContents=true!toolbox"),
```
What does `!toolbox` mean?

msuntharesan replied:
Since we are going to be using [react-toolbox](http://react-toolbox.com/) I was going to include [toolbox-loader](https://github.com/react-toolbox/toolbox-loader) for customizing. So should I still include it?

sokra replied:
just wondering, because it's not in the package.json

sokra commented on webpack.config.js:
```diff
@@ -0,0 +1,64 @@
+    loaders: [
+      {
+        test: /(\.jsx?)$/,
+        exclude: /node_modules/,
```
Please use `include` instead of `exclude`.

msuntharesan replied:
So we want to include `node_modules` as well?

alajmo replied:
No, I think he means, include what folder / files to use, instead of excluding files which you do not want to use:

```
include: [
  path.resolve(__dirname, "src")
],
```

This won't then go through node_modules.

msuntharesan replied:
@samiralajmovic Thank you.

alajmo commented:
@Cryptolemming my guess is because AirBnB is one of the biggest react users (they have a styleguide for it) and we'll be using react a lot. I also think it's fine that people make choices at this stage so that we can get started as quickly as possible (obviously if a lot ppl caution of using something we can halt for a bit and think about it). These are just my thoughts though.

abd124abd commented:
yeah agreed was just wondering - decisions are a good thing certainly.

learned that AirBnB is one of the biggest React users - thanks.

dlmr commented on .eslintrc:
```diff
@@ -1,37 +1,38 @@
+    "nodeca",
+    "react"
+  ],
+  "rules": {
```
Do we really need custom rules here? Everything should be included from `airbnb` I think.

Commit:
Webpack config fixes

sokra commented on webpack.config.test.js:
```diff
@@ -0,0 +1,41 @@
+const webpack = require("webpack");
+const autoprefixer = require("autoprefixer");
```
require "path" is missing here

sokra commented:
Is it possible to write the webpack config file in ES6 (with `import`)? For raw webpack you can rename them `webpack.config.babel.js`, but I'm not sure if this works with karma.

timaschew commented:
@jnwng the link to the styleguide is little bit confusing.
Seems that [http://airbnb.io/javascript/](http://airbnb.io/javascript/) is outdated, it doesn't contain anything about ES6 or React. But the link to the github repo does: [https://github.com/airbnb/javascript](https://github.com/airbnb/javascript)

Do we want to follow the whole styleguide? Some rules sounds crazy, for instance to use always semicolons.

msuntharesan commented:
> Do we want to follow the whole styleguide? Some rules sounds crazy, for instance to use always semicolons.

@timaschew Yes I agree. The dangling coma is another one

abd124abd commented:
not to extend this into a syntax discussion there are plenty out there - but semicolons and the like [still TC39 convention](https://blogs.windows.com/msedgedev/2015/08/03/recapping-the-july-2015-tc39-committee-meeting-in-redmond) and supported by Crockford etc

Commit:
Add missing require

msuntharesan commented:
So Plan of action for this PR
- switch to Babel 5
- use `babel-plugin-react-transform` and `react-transform-hmr`
- switch to `mocha` and `chai` perhaps?

jnwng commented:
@vanthiyathevan yep, those are good things to address.
@timaschew thanks for pointing it out - i've updated the link (i got lazy googling for it and found the wrong page). 

i dont think we need to follow the entire styleguide - the `eslintrc` is made extensible for us to customize to what makes sense for us. they _do_ however provide sane defaults for a majority of rules that most of us don't care about and just want to be consistent about, so i figure its a good starting point and the customization can come as needed.

zaggy commented:
Not intending to bring a syntax discussion, but my personal preferences are in favor of the entire airbnb styleguide. Especially dangling commas)) Oh, they are so useful and make merging so much easier... But I am up to anything the majority agrees too)

Commit:
Switching to babel 5

- using babel-plugin-react-transform and react-transform-hmr
- using mocha and chai

Commit:
Clean up npm scripts

use `[cross-env](https://github.com/kentcdodds/cross-env)` to set `NODE_ENV`

p0wl commented:
is anyone still working on this?
